File: Algorithms/Dynamic-Programming/LIS/LIS.go
package main

import "fmt"

func binarySearch(array []int, l int, r int, key int) int {
	for r-l > 1 {
		center := l + (r-l)/2
		if array[center] >= key {
			r = center
		} else {
			l = center
		}
	}

	return r
}

// LongestIncreasingSubsequence will returned a maximum increasing subsequence inside the input given
// with complexity O(N log N)
func LongestIncreasingSubsequence(input []int) int {
	if len(input) == 0 {
		return 0
	}

	length := 1
	array := make([]int, len(input))
	array[0] = input[0]

	for i := 1; i < len(array); i++ {
		if input[i] < array[0] {
			array[0] = input[i]
		} else if input[i] > array[length-1] {
			array[length] = input[i]
			length++
		} else {
			array[binarySearch(array, -1, length-1, input[i])] = input[i]
		}
	}

	return length
}

func main() {
	input := []int{1, 4, 3, 8, 12, 1, 9, 10, 3, 11, 0, 3}
	output := LongestIncreasingSubsequence(input)

	fmt.Println(output)
}

File: Algorithms/Dynamic-Programming/LIS/LIS.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Input Format: length of the array and contents of the array
Output Format: length of LIS and contents of the LIS
ALgorithm: Bottom Up Dynamic programming
Time Complexity: O(N^2), where N is the length of the array
Space Complexity: O(N^2)

Sample Input and Output

**Sample Input 1:**

Enter the number of elements of the array: 8

Enter the elements of the array: 1 8 5 9 6 2 4 7

The length of LIS is: 4

The LIS is: 1 5 6 7

**Sample Input 2:**

Enter the number of elements of the array: 10

Enter the elements of the array: 1 2 3 6 5 8 9 4 7 6

The length of LIS is: 6

The LIS is: 1 2 3 6 8 9

Graph contains cycle!
*/

int main()
{

  int n; // number of elements of the array
  cout << "\nEnter the number of elements of the array: ";
  cin >> n;

  vector<int> a(n);

  vector<int> lis(n); // for storing the longest increasing subsequence size of each index

  vector<vector<int>> seq(n); // storing the longest increasing subsequence for each index from 0 to n-1

  cout << "\nEnter the elements of the array: ";
  for (int i = 0; i < n; i++)
  {
    cin >> a[i];
  }

  // initialising the LIS for each index to 1
  for (int i = 0; i < n; i++)
  {
    lis[i] = 1;
  }

  // counting the LIS in bottom up manner dynamic programming
  for (int i = 1; i < n; i++)
  {
    vector<int> tmp;
    for (int j = 0; j < i; j++)
    {
      // if the current LIS value is less that the value we just calculated we have got a new LIS for index i
      if (a[i] > a[j] and lis[i] < lis[j] + 1)
      {
        lis[i] = lis[j] + 1;
        tmp.push_back(a[j]);
      }
    }
    tmp.push_back(a[i]);
    seq[i] = tmp;
  }

  int answerIndex = 0;
  int maxLength = *max_element(lis.begin(), lis.end());
  for (int i = 0; i < n; i++)
  {
    if ((int)seq[i].size() == maxLength)
    {
      answerIndex = i;
      break;
    }
  }
  // printing the LIS and contents of the LIS
  cout << "\nThe length of LIS is: " << maxLength << '\n';
  cout << "\nThe LIS is: ";
  for (auto u : seq[answerIndex])
  {
    cout << u << " ";
  }
  cout << '\n';
}